static-analysis-plugin/plugin/src/test/groovy/com/novoda/test/TestProjectRule.groovy

```diff
@@ -0,0 +1,40 @@
+package com.novoda.test
+
+import org.junit.rules.TestRule
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+
+final class TestProjectRule implements TestRule {
+
+    private final Closure<TestProject> projectFactory
+    private TestProject project
+
+    static TestProjectRule forJavaProject() {
+        new TestProjectRule({ new TestJavaProject() })
+    }
+
+    static TestProjectRule forAndroidProject() {
+        new TestProjectRule({ new TestAndroidProject() })
+    }
+
+    private TestProjectRule(Closure<TestProject> projectFactory) {
+        this.projectFactory = projectFactory
+    }
+
+    public TestProject newProject() {
+        project = projectFactory.call()
+        return project
+    }
+
+    @Override
+    Statement apply(Statement base, Description description) {
+        return new Statement() {
+            @Override
+            void evaluate() throws Throwable {
+                base.evaluate()
+                project?.deleteDir()
```

Comment:
test project folders can be created invoking `newProject()` on the test rule, and if the test is successful they will be deleted at the end of each test